core/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java

```diff
@@ -0,0 +1,117 @@
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.IndexService;
+import org.elasticsearch.index.shard.IndexShard;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.indices.IndicesService;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+
+import java.io.IOException;
+
+public class GlobalCheckpointSyncAction extends TransportReplicationAction<ReplicationRequest, GlobalCheckpointSyncAction.ReplicaRequest, ReplicationResponse> {
+
+    public static String ACTION_NAME = "indices:admin/seq_no/global_checkpoint_sync";
+
+    @Inject
+    public GlobalCheckpointSyncAction(Settings settings, TransportService transportService, ClusterService clusterService,
+                                      IndicesService indicesService, ThreadPool threadPool, ShardStateAction shardStateAction,
+                                      MappingUpdatedAction mappingUpdatedAction,
+                                      ActionFilters actionFilters, IndexNameExpressionResolver indexNameExpressionResolver) {
+        super(settings, ACTION_NAME, transportService, clusterService, indicesService, threadPool, shardStateAction, mappingUpdatedAction,
+            actionFilters, indexNameExpressionResolver, ReplicationRequest::new, ReplicaRequest::new, ThreadPool.Names.SAME);
+    }
+
+    @Override
+    protected ReplicationResponse newResponseInstance() {
+        return new ReplicationResponse();
+    }
+
+    @Override
+    protected Tuple<ReplicationResponse, ReplicaRequest> shardOperationOnPrimary(MetaData metaData, ReplicationRequest request) throws Throwable {
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(request.shardId().id());
+        long checkPoint = indexShard.getGlobalCheckpoint();
+        return new Tuple<>(new ReplicationResponse(), new ReplicaRequest(request, checkPoint));
+    }
+
+    @Override
+    protected void shardOperationOnReplica(ReplicaRequest request) {
+        IndexService indexService = indicesService.indexServiceSafe(request.shardId().getIndex());
+        IndexShard indexShard = indexService.getShard(request.shardId().id());
+        indexShard.updateGlobalCheckpointOnReplica(request.checkPoint);
+    }
+
+    public void updateCheckpointForShard(ShardId shardId) {
+        execute(new ReplicationRequest(shardId), new ActionListener<ReplicationResponse>() {
+            @Override
+            public void onResponse(ReplicationResponse replicationResponse) {
+                logger.trace("{} global check point successfully updated (shard info [{}])", shardId, replicationResponse.getShardInfo());
```

Comment:
Since this is in the common code path and because of `ShardInfo#toString` using `XContent` to generate its value, I think this should be wrapped in a `logger#isTraceEnabled` check.